File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        
        intervals.sort(key=lambda x:x[0])
        
        ans = [intervals[0]]
        
        for start, end in intervals[1:]:
            last_start, last_end = ans[-1]
            
            if last_end >= start:
                ans[-1] = [
                    last_start,
                    max(last_end, end)
                ]
            else:
                ans.append([start, end])
        
        return ans
        

File: 0056-merge-intervals/0056-merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    // sort intervals by starting point
    let sorted = intervals.sort((a,b) => a[0] - b[0])
    
    // keep track of curr element
    // if current overlaps with next element
    // make new element where start is smallest and end is biggest
    
    // curr start is equal to next start they overlap, new element start = value, new element end = greater value
    // curr end is greater than start next 
    let ans = []
    let i = 0
    while (sorted[i]) {
        let curr = sorted[i]
        let next = sorted[i + 1]
        
        
        if (curr[1] >= next?.[0]) {
            let min = curr[0]
            let max = Math.max(curr[1], next[1])
            sorted.splice(i, 2, [min, max])
                    
        } else {
            i++
        }
    
        
    }
    return sorted
};